prow/cmd/deck/main.go

```diff
@@ -1663,3 +1670,40 @@ func defaultLensRemoteConfig(lfc *config.LensFileConfig) error {
 
 	return nil
 }
+
+// ValidatePath validates various parts (e.g. bucket, folder, etc.) of a storage path.
+func ValidatePath(cfg config.Getter, path string) error {
+	if !cfg().Deck.EnableWhitelist {
+		return nil
+	}
+
+	parts := strings.SplitN(path, "/", 4)
+	if len(parts) < 4 {
+		return fmt.Errorf("invalid path: %s expected <key-type>/<bucket>/<folder>/<sub-folder...>", path)
+	}
+
+	// Check bucket
+	bucket := parts[1]
+	bwl := sets.String{}
+	for _, dc := range cfg().Plank.DefaultDecorationConfigs {
+		bwl.Insert(dc.GCSConfiguration.Bucket)
+	}
+	for _, bucket := range cfg().Deck.AdditionalBuckets {
+		bwl.Insert(bucket)
+	}
+	if !bwl.Has(bucket) {
+		return fmt.Errorf("bucket %q not in whitelist %v", bucket, bwl)
+	}
+
+	// Check folder
+	folder := parts[2]
+	fwl := sets.NewString(gcs.PRLogs, gcs.NonPRLogs)
+	for _, folder := range cfg().Deck.AdditionalFolders {
+		fwl.Insert(folder)
```

Comment:
nit: The folder whitelist is currently limited to whitelisting top-level folders. It could be made more flexibly by matching against the full path (with a prefix or regexp match) to allow specific sub folders to be whitelisted.
I don't think this matters much though.

Reply:
I went with whitelisting of top-level folders because it is the most performant and simplest way to gate access. If prefix or regex matching is a need that arises then I can add it.